utils/generate-command-code.py

```diff
@@ -0,0 +1,489 @@
+            for subcommand in self.subcommands:
+                f.write("{%s},\n" % subcommand.struct_code())
+            f.write("{0}\n")
+            f.write("};\n\n")
+
+        f.write("/********** %s ********************/\n\n" % self.fullname())
+
+        f.write("/* %s return info */\n" % self.fullname())
+        code = self.return_info_code()
+        if code:
+            f.write("commandReturnInfo %s[] = {\n" % self.return_types_table_name())
+            f.write("%s\n" % code)
+            f.write("};\n\n")
+        else:
+            f.write("#define %s NULL\n\n" % self.return_types_table_name())
+
+        f.write("/* %s history */\n" % self.fullname())
+        code = self.history_code()
+        if code:
+            f.write("commandHistory %s[] = {\n" % self.history_table_name())
+            f.write("%s\n" % code)
+            f.write("};\n\n")
+        else:
+            f.write("#define %s NULL\n\n" % self.history_table_name())
+
+        if self.args:
+            for arg in self.args:
+                arg.write_internal_structs(f)
+
+            f.write("/* %s argument table */\n" % self.fullname())
+            f.write("struct redisCommandArg %s[] = {\n" % self.arg_table_name())
+            for arg in self.args:
+                f.write("{%s},\n" % arg.struct_code())
+            f.write("{0}\n")
+            f.write("};\n\n")
+
+        #f.write("/* %s command */\n" % self.fullname())
+        #f.write("#define %s {%s}\n\n" % (self.struct_name(), self.struct_code()))
+
+
+class Subcommand(Command):
+    def __init__(self, name, desc):
+        self.container_name = desc["container"].upper()
+        super(Subcommand, self).__init__(name, desc)
+
+    def fullname(self):
+        return "%s %s" % (self.container_name, self.name)
+
+
+def create_command(name, desc):
+    if desc.get("container"):
+        cmd = Subcommand(name.upper(), desc)
+        subcommands.setdefault(desc["container"].upper(), {})[name] = cmd
+    else:
+        cmd = Command(name.upper(), desc)
+        commands[name.upper()] = cmd
+
+
+# MAIN
+
+# Create all command objects
+print("Processing json files...")
+for filename in glob.glob('commands/*.json'):
+    print(filename)
+    with open(filename,"r") as f:
+        d = json.load(f)
+        for name, desc in d.items():
+            create_command(name, desc)
+
+# Link subcommands to containers
+print("Linking container command to subcommands...")
+for command in commands.values():
+    if command.name not in subcommands:
+        continue
+    for subcommand in subcommands[command.name].values():
+        command.subcommands.append(subcommand)
+
+print("Generating commands.c...")
+with open("commands.c","w") as f:
+    f.write("/* Automatically generated by %s, do not edit. */\n\n" % os.path.basename(__file__))
+    f.write("#include \"server.h\"\n")
+    f.write(
+"""
+/* Our command table.
+*
+* (See comment above sturct redisCommand)
```

Comment:
re-thinking that huge doc-comment again, if we keep it here in this form, we'll have it twice in our repo.
i suppose there's no sense in having it next to the generated code, and there's probably no sense having it in the generator script.
the best place to have it would be next to the "code" we keep editing (like we had in the past, where it was next to the command table), so now that would be the individual json files, where we obviously don't want it.
so let's figure out where we would look for the doc when we edit the json files.
it's probably not commands.c or the generator script..
maybe move to server.h next to the redisCommand struct?